pkg/apis/monitoring/v1/servicemonitor_types.go

```diff
@@ -83,6 +85,13 @@ type ServiceMonitorSpec struct {
 
 	// Label selector to select the Kubernetes `Endpoints` objects to scrape metrics from.
 	Selector metav1.LabelSelector `json:"selector"`
+
+	// Role selector to select the Kubernetes `Endpoints` objects to scrape metrics from.
+	// If `roleSelector` is defined, the `selector` field is ignored.
```

Comment:
I feel that the API doesn't provide a super-clean user experience.

Say that you have

````
spec:
  selector:
    matchLabels: {foo: bar}
````

If you want to use role selectors instead, you could change it to the following but it requires to do the translation "by hand".

````
spec:
  roleSelector:
  - role: Service
    labels:
    - key: foo
      operator: In
      values: [bar]
  - role: Endpoints
    labels:
    - key: foo
      operator: In
      values: [bar]
````

But you can also define bogus configs like:

````
spec:
  selector:
    matchLabels: {foo: bar}
  # because role selector list isn't empty, the selector field is completely ignored
  roleSelector:
  - role: Pod
    labels:
    - key: other_label
      operator: In
      values: [a,b,c]
````

Which is why I advocated for an enum field which would tell the operator to use role selectors instead of relabeling config to filter the targets: it would probably be enough for most users (and less error prone). And we can still expose role selectors for other resources.